util/proxy_protocol.c

```diff
@@ -0,0 +1,194 @@
+	{ PP_PARSE_WRONG_HEADERv2, "could not match PROXYv2 header" },
+	{ PP_PARSE_UNKNOWN_CMD, "unknown command" },
+	{ PP_PARSE_UNKNOWN_FAM_PROT, "unknown family and protocol" },
+};
+
+void
+pp_init(void (*write_uint16)(void* buf, uint16_t data),
+	void (*write_uint32)(void* buf, uint32_t data)) {
+	pp_data.write_uint16 = write_uint16;
+	pp_data.write_uint32 = write_uint32;
+}
+
+const char*
+pp_lookup_error(enum pp_parse_errors error) {
+	return pp_parse_errors_data[error].text;
+}
+
+size_t
+pp2_write_to_buf(uint8_t* buf, size_t buflen,
+#ifdef INET6
+	struct sockaddr_storage* src,
+#else
+	struct sockaddr_in* src,
+#endif
+	int stream)
+{
+	int af;
+	size_t expected_size;
+	if(!src) return 0;
+	af = (int)((struct sockaddr_in*)src)->sin_family;
+	expected_size = PP2_HEADER_SIZE + (af==AF_INET?12:36);
+	if(buflen < expected_size) {
+		return 0;
+	}
+	/* sig */
+	memcpy(buf, PP2_SIG, PP2_SIG_LEN);
+	buf += PP2_SIG_LEN;
+	/* version and command */
+	*buf = (PP2_VERSION << 4) | PP2_CMD_PROXY;
+	buf++;
+	if(af==AF_INET) {
+		/* family and protocol */
+		*buf = (PP2_AF_INET<<4) |
+			(stream?PP2_PROT_STREAM:PP2_PROT_DGRAM);
+		buf++;
+		/* length */
+		(*pp_data.write_uint16)(buf, 12);
+		buf += 2;
+		/* src addr */
+		memcpy(buf,
+			&((struct sockaddr_in*)src)->sin_addr.s_addr, 4);
+		buf += 4;
+		/* dst addr */
+		(*pp_data.write_uint32)(buf, 0);
+		buf += 4;
+		/* src port */
+		memcpy(buf,
+			&((struct sockaddr_in*)src)->sin_port, 2);
+		buf += 2;
+		/* dst addr */
+		/* dst port */
+		(*pp_data.write_uint16)(buf, 12);
+	} else {
+#ifdef INET6
+		/* family and protocol */
+		*buf = (PP2_AF_INET6<<4) |
+			(stream?PP2_PROT_STREAM:PP2_PROT_DGRAM);
+		buf++;
+		/* length */
+		(*pp_data.write_uint16)(buf, 36);
+		buf += 2;
+		/* src addr */
+		memcpy(buf,
+			&((struct sockaddr_in6*)src)->sin6_addr, 16);
+		buf += 16;
+		/* dst addr */
+		memset(buf, 0, 16);
+		buf += 16;
+		/* src port */
+		memcpy(buf, &((struct sockaddr_in6*)src)->sin6_port, 2);
+		buf += 2;
+		/* dst port */
+		(*pp_data.write_uint16)(buf, 0);
+#else
+		return 0;
+#endif /* INET6 */
+	}
+	return expected_size;
+}
+
+int
+pp2_read_header(uint8_t* buf, size_t buflen)
+{
+	size_t size;
+	struct pp2_header* header = (struct pp2_header*)buf;
+	/* Try to fail all the unsupported cases first. */
+	if(buflen < PP2_HEADER_SIZE) {
+		return PP_PARSE_SIZE;
+	}
+	/* Check for PROXYv2 header */
+	if(memcmp(header, PP2_SIG, PP2_SIG_LEN) != 0 ||
+		((header->ver_cmd & 0xF0)>>4) != PP2_VERSION) {
+		return PP_PARSE_WRONG_HEADERv2;
+	}
+	/* Check the length */
+	size = PP2_HEADER_SIZE + ntohs(header->len);
+	if(buflen < size) {
+		return PP_PARSE_SIZE;
+	}
+	/* Check for supported commands */
+	if((header->ver_cmd & 0xF) != PP2_CMD_LOCAL &&
+		(header->ver_cmd & 0xF) != PP2_CMD_PROXY) {
+		return PP_PARSE_UNKNOWN_CMD;
+	}
+	/* Check for supported family and protocol */
+	if(header->fam_prot != 0x00 /* AF_UNSPEC|UNSPEC */ &&
+		header->fam_prot != 0x11 /* AF_INET|STREAM */ &&
```

Comment:
This code, util/proxy_protocol.c is from the generic proxy protocol code, that is in Unbound.

The constants for it as defined in util/proxy_protocol.h. I think it was chosen to display like this to be more readable, than the bit munging that is needed with the constants, because the values align with hex characters. That said, I am fine with this code but also with a use of the defined constants.

Reply:
Defined constants should work better since I see that these combinations are also used elsewhere in the code.

Reply:
Resolved in d8a47b171ca92092af8c33a888487726f72c77b4.